flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/internal/converter/AbstractJdbcRowConverter.java

```diff
@@ -18,54 +18,255 @@
+			toExternalConverters[i] = createNullableExternalConverter(fieldTypes[i]);
 		}
 	}
 
 	@Override
-	public Row convert(ResultSet resultSet, Row reuse) throws SQLException {
+	public RowData toInternal(ResultSet resultSet) throws SQLException {
+		GenericRowData genericRowData = new GenericRowData(rowType.getFieldCount());
 		for (int pos = 0; pos < rowType.getFieldCount(); pos++) {
-			reuse.setField(pos, converters[pos].convert(resultSet.getObject(pos + 1)));
+			genericRowData.setField(pos, toInternalConverters[pos].deserialize(resultSet.getObject(pos + 1)));
+		}
+		return genericRowData;
+	}
+
+	@Override
+	public PreparedStatement toExternal(RowData rowData, PreparedStatement statement) throws SQLException {
+		for (int index = 0; index < rowData.getArity(); index++) {
+			toExternalConverters[index].serialize(rowData, index, statement);
+		}
+		return statement;
+	}
+
+	/**
+	 * Runtime converter to convert JDBC field to {@link RowData} type object.
+	 */
+	@FunctionalInterface
+	interface JdbcDeserializationConverter extends Serializable {
+		/**
+		 * convert a jdbc field to java object, the field could be a simple type or array type.
+		 * @param jdbcField
+		 */
+		Object deserialize(Object jdbcField) throws SQLException;
+	}
+
+	/**
+	 * Runtime converter to convert {@link RowData} field to java object and fill into the {@link PreparedStatement}.
+	 */
+	@FunctionalInterface
+	interface JdbcSerializationConverter extends Serializable {
+		void serialize(RowData rowData, int index, PreparedStatement statement) throws SQLException;
+	}
+
+	/**
+	 * Create a nullable runtime {@link JdbcDeserializationConverter} from given {@link LogicalType}.
+	 */
+	protected JdbcDeserializationConverter createNullableInternalConverter(LogicalType type) {
+		return wrapIntoNullableInternalConverter(createInternalConverter(type));
+	}
+
+	protected JdbcDeserializationConverter wrapIntoNullableInternalConverter(JdbcDeserializationConverter jdbcDeserializationConverter) {
+		return v -> {
+			if (v == null) {
+				return null;
+			} else {
+				return jdbcDeserializationConverter.deserialize(v);
+			}
+		};
+	}
+
+	protected JdbcDeserializationConverter createInternalConverter(LogicalType type) {
+		switch (type.getTypeRoot()) {
+			case NULL:
+				return v -> null;
+			case BOOLEAN:
+				return v -> (boolean) v;
+			case TINYINT:
+				return v -> (byte) v;
+			case SMALLINT:
+				// Converter for small type that casts value to int and then return short value, since
+				// JDBC 1.0 use int type for small values.
+				return v -> (Integer.valueOf(v.toString())).shortValue();
+			case INTEGER:
+			case INTERVAL_YEAR_MONTH:
+				return v -> (int) v;
+			case BIGINT:
+			case INTERVAL_DAY_TIME:
+				return v -> (long) v;
+			case DATE:
+				return v -> (int) (((Date) v).toLocalDate().toEpochDay());
+			case TIME_WITHOUT_TIME_ZONE:
+				return v -> ((Time) v).toLocalTime().toSecondOfDay() * 1000;
```

Comment:
Would be better to use `return v -> ((Time) v).toLocalTime().toNanoOfDay() / 1_000_000L;` to support millisecond.